Task nxdoc 2696 2023 to 2025 upgrade (#2381)
nuxeo/doc.nuxeo.com-content: src/nxdoc/nuxeo-server/upgrading-the-nuxeo-platform/how-to-upgrade-from-lts-2023-to-lts-2025/how-to-upgrade-to-jakarta-rs-3.md

---
title: How to upgrade to Jakarta RS 3 - WebEngine
description: Instruction to upgrade your Nuxeo Project to LTS 2025 - Jakarta RS part
review:
   comment: ''
   date: '2025-03-12'
   status: ok
labels:
    - release-notes
toc: true
tree_item_index: 200
---

## Introduction

{{! excerpt}}
While the upgrade to **Jakarta RS 3** is part of the Jakarta EE 10 one, it has its own section because the implementation

> (Jersey), whose Nuxeo WebEngine is leveraging, has more changes.

Reviewer: ```suggestion
(Jersey), leveraged by Nuxeo WebEngine, has more changes.
```

> Nuxeo WebEngine itself has received breaking changes to benefit from the new Jakarta RS additions.

Reviewer: ```suggestion
Nuxeo WebEngine has undergone breaking changes to benefit from the new Jakarta RS additions.
```

{{! /excerpt}}

You must first follow the [How to upgrade to Jakarta EE 10]({{page page='how-to-upgrade-to-jakarta-ee-10'}}) page.

> _JAX-RS_ occurrences have been removed from the Nuxeo Server code, it has been renamed to _REST_ where it makes sense.

Reviewer: ```suggestion
 _JAX-RS_ occurrences have been replaced by  _REST_ within the Nuxeo Server codebase when relevant.
```

## Maven Dependency

In addition to the Jakarta RS Specification and Implementation dependencies to change, you will need to update:
- `nuxeo-webengine-jaxrs` to `nuxeo-webengine-rest`
- `nuxeo-importer-jaxrs` to `nuxeo-importer-rest`
- `nuxeo-target-platforms-jaxrs` to `nuxeo-target-platforms-rest`
- `nuxeo-template-rendering-jaxrs` to `nuxeo-template-rendering-rest`

## Java Code

One of the biggest change in the Jersey Implementation is the total rework of their Dependency Injection engine.
It is now more extendable, and better fits with the direction of Jakarta RS specification to support `jakarta.inject-api`
annotations, such as `@Singleton` which is currently used.

### Package

> The following Nuxeo packages have changed in order to get rid of _JAX-RS_ occurrences:

Reviewer: `get rid` is a bit too familiar, isn't it?

Author: Sounds okay to me.

- `org.nuxeo.ecm.webengine.jaxrs` to `org.nuxeo.ecm.webengine.rest`
- `org.nuxeo.ecm.restapi.server.jaxrs` to `org.nuxeo.ecm.restapi.server`
- `org.nuxeo.ecm.automation.jaxrs.io` to `org.nuxeo.ecm.automation.io.rest`
- `org.nuxeo.ecm.automation.server.jaxrs` to `org.nuxeo.ecm.automation.server.rest`
- `org.nuxeo.targetplatforms.jaxrs` to `org.nuxeo.targetplatforms.rest`
- `org.nuxeo.template.jaxrs` to `org.nuxeo.template.rest`

> In the case you were extending the Nuxeo REST API, you must also update the package `org.nuxeo.ecm.restapi.server.jaxrs` 
> in your project to `org.nuxeo.ecm.restapi.server` to make the Package Scan work correctly.

Reviewer: ```suggestion
If you were extending the Nuxeo REST API, your package `org.nuxeo.ecm.restapi.server.jaxrs` must be updated to `org.nuxeo.ecm.restapi.server` to keep the Package Scan working.
```

### Nuxeo Component

> The following Nuxeo bundle names have changed in order to get rid of _JAX-RS_ occurrences:

Reviewer: `get rid` is too informal

- `org.nuxeo.ecm.webengine.jaxrs` to `org.nuxeo.ecm.webengine.rest`
- `org.nuxeo.targetplatforms.jaxrs` to `org.nuxeo.targetplatforms.rest`
- `org.nuxeo.template.manager.jaxrs` to `org.nuxeo.template.manager.rest`

> You must update their occurrences in `@Deploy` annotations in tests.

Reviewer: ```suggestion
Test `@Deploy` annotation references must be updated accordingly.
```

### Message Body Writer Selection Algorithm

> The way the `MessageBodyWriter` is chosen has changed in this upgrade, you can read more in the [Jakarta Documentation](https://eclipse-ee4j.github.io/jersey.github.io/documentation/latest/message-body-workers.html#providers-selection).

Reviewer: do you mean "is retrieved" or "is selected"?

Author: I would have say both, it is selected then retrieved to marshall the response.

> The way the `MessageBodyWriter` is chosen has changed in this upgrade, you can read more in the [Jakarta Documentation](https://eclipse-ee4j.github.io/jersey.github.io/documentation/latest/message-body-workers.html#providers-selection).

Reviewer: ```suggestion
The `MessageBodyWriter` is retrieved differently in this upgrade as explained in the [Jakarta Documentation](https://eclipse-ee4j.github.io/jersey.github.io/documentation/latest/message-body-workers.html#providers-selection).
```

    at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:163) ~[jersey-common-2.27.jar:?]
    at org.glassfish.jersey.server.internal.MappableExceptionWrapperInterceptor.aroundWriteTo(MappableExceptionWrapperInterceptor.java:85) ~[jersey-server-2.27.jar:?]
    ... 66 more
```

```
Caused by: javax.ws.rs.NotAcceptableException: HTTP 406 Not Acceptable
    at org.glassfish.jersey.server.internal.routing.MethodSelectingRouter.getMethodRouter(MethodSelectingRouter.java:471)
    at org.glassfish.jersey.server.internal.routing.MethodSelectingRouter.access$000(MethodSelectingRouter.java:73)
    at org.glassfish.jersey.server.internal.routing.MethodSelectingRouter$4.apply(MethodSelectingRouter.java:673)
    at org.glassfish.jersey.server.internal.routing.MethodSelectingRouter.apply(MethodSelectingRouter.java:304)
    at org.glassfish.jersey.server.internal.routing.RoutingStage._apply(RoutingStage.java:86)
    at org.glassfish.jersey.server.internal.routing.RoutingStage._apply(RoutingStage.java:89)
```

In such cases you should declare the concrete type: `DocoumentModel`.

> If the method returns different incompatible type, you can replace Object by `jakarta.ws.rs.core.Response` (and adapt 
> the code to return a `Response` with help of `ResponseBuilder`), in this case Jersey won't leverage the method 
> reflection introspection.

Reviewer: ```suggestion
If the method returns a different incompatible type, you can replace Object by `jakarta.ws.rs.core.Response` (and adapt the code to return a `Response` with the help of a `ResponseBuilder`), that Jersey doesn't leverage the method reflection introspection.
```

### WebContext Managed by A Jakarta Provider

> Previously, the `WebContext` was initialized with a Servlet `Filter`, and filled latter on with its `HttpHeaders` and `UriInfo`.

Reviewer: `s/latter/later`

Now in LTS 2025, the `WebContext` is initialiazed by a Jakarta `Provider` and immutable. This allows to access the context

> in any Jakarta objects with help of `@Context` annotation.

Reviewer: ```suggestion
in any Jakarta objects with the help of the `@Context` annotation.
```

For example, the following code:

```java
    @POST
    public Response doPost() {
        DocumentModel doc = getTarget().getAdapter(DocumentModel.class);
        FormData form = ctx.getForm();
        form.fillDocument(doc);
        String xpath = ctx.getForm().getString(FormData.PROPERTY);
        Blob blob = form.getFirstBlob();
        if (xpath == null || blob == null) {
            throw new RuntimeException("Illegal parameters");
        }
        ...
    }
```

Should be updated to:

```java
import javax.ws.rs.FormParam;
import javax.ws.rs.core.MultivaluedMap;
import org.glassfish.jersey.media.multipart.FormDataParam;

    @POST
    public Response doPost(MultivaluedMap<String, String> formParams, @FormParam("property") String xpath,
            @FormDataParam("content") Blob blob) {
        DocumentModel doc = getTarget().getAdapter(DocumentModel.class);
        DocumentHelper.fillDocument(doc, formParams);
        if (xpath == null || blob == null) {
            throw new RuntimeException("Illegal parameters");
        }
        ...
    }
```

The `@FormParam` annotations allows to retrieve the value of the named form parameter. If you need a dynamic access

> to form parameters, you should declare one parameter with `MultivaluedMap<String, String>` that will receive them.

Reviewer: ```suggestion
The `@FormParam` annotations allow retrieving the value of the named form parameter. If you need dynamic access to form parameters, you should declare a parameter with `MultivaluedMap<String, String>` that will collect them.
```

> The `@FormParam` annotations allows to retrieve the value of the named form parameter. If you need a dynamic access
> to form parameters, you should declare one parameter with `MultivaluedMap<String, String>` that will receive them.

Reviewer: ```suggestion
The `@FormParam` annotations allow retrieving the value of the named form parameter. If you need dynamic access to form parameters, you should declare a parameter with `MultivaluedMap<String, String>` that will collect them.
```

### ModuleRoot Does Not Hold HttpServletRequest Anymore

The `HttpServletRequest request` field present in `ModuleRoot` has been removed as it is not needed anymore since
`WebContext` can be retrieved from the Jakarta Context.

> If you need it just declare it as a field or method parameter in your code:

Reviewer 2: ```suggestion
If you need it, just declare it as a field or method parameter in your code:
```

---
title: How to upgrade to Jakarta RS 3 - WebEngine

> description: Instruction to upgrade your Nuxeo Project to LTS 2025 - Jakarta RS part

Reviewer 3: ```suggestion
description: Instructions to upgrade your Nuxeo Project to LTS 2025 - Jakarta RS part
```

- `nuxeo-webengine-jaxrs` to `nuxeo-webengine-rest`
- `nuxeo-importer-jaxrs` to `nuxeo-importer-rest`
- `nuxeo-target-platforms-jaxrs` to `nuxeo-target-platforms-rest`
- `nuxeo-template-rendering-jaxrs` to `nuxeo-template-rendering-rest`

## Java Code

One of the biggest change in the Jersey Implementation is the total rework of their Dependency Injection engine.
It is now more extendable, and better fits with the direction of Jakarta RS specification to support `jakarta.inject-api`
annotations, such as `@Singleton` which is currently used.

### Package

The following Nuxeo packages have changed in order to get rid of _JAX-RS_ occurrences:
- `org.nuxeo.ecm.webengine.jaxrs` to `org.nuxeo.ecm.webengine.rest`
- `org.nuxeo.ecm.restapi.server.jaxrs` to `org.nuxeo.ecm.restapi.server`
- `org.nuxeo.ecm.automation.jaxrs.io` to `org.nuxeo.ecm.automation.io.rest`
- `org.nuxeo.ecm.automation.server.jaxrs` to `org.nuxeo.ecm.automation.server.rest`
- `org.nuxeo.targetplatforms.jaxrs` to `org.nuxeo.targetplatforms.rest`
- `org.nuxeo.template.jaxrs` to `org.nuxeo.template.rest`

If you were extending the Nuxeo REST API, your package `org.nuxeo.ecm.restapi.server.jaxrs` must be updated to 

> `org.nuxeo.ecm.restapi.server` to keep the Package Scan working.

Reviewer 3: Maybe wrap it in a "warning" section or something?

    at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:163) ~[jersey-common-2.27.jar:?]
    at org.glassfish.jersey.server.internal.MappableExceptionWrapperInterceptor.aroundWriteTo(MappableExceptionWrapperInterceptor.java:85) ~[jersey-server-2.27.jar:?]
    ... 66 more
```

```
Caused by: javax.ws.rs.NotAcceptableException: HTTP 406 Not Acceptable
    at org.glassfish.jersey.server.internal.routing.MethodSelectingRouter.getMethodRouter(MethodSelectingRouter.java:471)
    at org.glassfish.jersey.server.internal.routing.MethodSelectingRouter.access$000(MethodSelectingRouter.java:73)
    at org.glassfish.jersey.server.internal.routing.MethodSelectingRouter$4.apply(MethodSelectingRouter.java:673)
    at org.glassfish.jersey.server.internal.routing.MethodSelectingRouter.apply(MethodSelectingRouter.java:304)
    at org.glassfish.jersey.server.internal.routing.RoutingStage._apply(RoutingStage.java:86)
    at org.glassfish.jersey.server.internal.routing.RoutingStage._apply(RoutingStage.java:89)
```

In such cases you should declare the concrete type: `DocoumentModel`.

> If the method returns different incompatible types, you can replace Object by `jakarta.ws.rs.core.Response` (and adapt 
> the code to return a `Response` with help of `ResponseBuilder`), that Jersey doesn't leverage the method reflection introspection.

Reviewer 3: ```suggestion
If the method returns different incompatible types, you can replace `Object` by `jakarta.ws.rs.core.Response` (and adapt 
the code to return a `Response` with help of `ResponseBuilder`). This ensures that Jersey does not rely on method reflection introspection.
```

### WebContext Managed by A Jakarta Provider

Previously, the `WebContext` was initialized with a Servlet `Filter`, and filled later on with its `HttpHeaders` and `UriInfo`.

> Now in LTS 2025, the `WebContext` is initialiazed by a Jakarta `Provider` and immutable. This allows to access the context

Reviewer 3: ```suggestion
Now in LTS 2025, the `WebContext` is initialiazed by a Jakarta `Provider` and is immutable. This allows to access the context
```

Should be updated to:

```java
import javax.ws.rs.FormParam;
import javax.ws.rs.core.MultivaluedMap;
import org.glassfish.jersey.media.multipart.FormDataParam;

    @POST
    public Response doPost(MultivaluedMap<String, String> formParams, @FormParam("property") String xpath,
            @FormDataParam("content") Blob blob) {
        DocumentModel doc = getTarget().getAdapter(DocumentModel.class);
        DocumentHelper.fillDocument(doc, formParams);
        if (xpath == null || blob == null) {
            throw new RuntimeException("Illegal parameters");
        }
        ...
    }
```

The `@FormParam` annotations allow retrieving the value of the named form parameter. If you need dynamic access
to form parameters, you should declare one parameter with `MultivaluedMap<String, String>` that will collect them.

### ModuleRoot Does Not Hold HttpServletRequest Anymore

The `HttpServletRequest request` field present in `ModuleRoot` has been removed as it is not needed anymore since
`WebContext` can be retrieved from the Jakarta Context.

If you need it, just declare it as a field or method parameter in your code:

```java

    @Context
    protected HttpServletRequest request;

```

> ### Web Engine Module Singletons

Reviewer 3: ```suggestion
### WebEngine Module Singletons
```

{{! excerpt}}
While the upgrade to **Jakarta RS 3** is part of the Jakarta EE 10 one, it has its own section because the implementation
(Jersey), leveraged by Nuxeo WebEngine, has more changes.

Nuxeo WebEngine has undergone breaking changes to benefit from the new Jakarta RS additions.
{{! /excerpt}}

You must first follow the [How to upgrade to Jakarta EE 10]({{page page='how-to-upgrade-to-jakarta-ee-10'}}) page.

_JAX-RS_ occurrences have been replaced by _REST_ within the Nuxeo Server codebase.

## Maven Dependency

In addition to the Jakarta RS Specification and Implementation dependencies to change, you will need to update:
- `nuxeo-webengine-jaxrs` to `nuxeo-webengine-rest`
- `nuxeo-importer-jaxrs` to `nuxeo-importer-rest`
- `nuxeo-target-platforms-jaxrs` to `nuxeo-target-platforms-rest`
- `nuxeo-template-rendering-jaxrs` to `nuxeo-template-rendering-rest`

## Java Code

One of the biggest change in the Jersey Implementation is the total rework of their Dependency Injection engine.
It is now more extendable, and better fits with the direction of Jakarta RS specification to support `jakarta.inject-api`
annotations, such as `@Singleton` which is currently used.

### Package

> The following Nuxeo packages have changed in order to get rid of _JAX-RS_ occurrences:

Reviewer 4: ```suggestion
The following Nuxeo packages have changed in order to remove _JAX-RS_ occurrences:
```

- `org.nuxeo.ecm.webengine.jaxrs` to `org.nuxeo.ecm.webengine.rest`
- `org.nuxeo.ecm.restapi.server.jaxrs` to `org.nuxeo.ecm.restapi.server`
- `org.nuxeo.ecm.automation.jaxrs.io` to `org.nuxeo.ecm.automation.io.rest`
- `org.nuxeo.ecm.automation.server.jaxrs` to `org.nuxeo.ecm.automation.server.rest`
- `org.nuxeo.targetplatforms.jaxrs` to `org.nuxeo.targetplatforms.rest`
- `org.nuxeo.template.jaxrs` to `org.nuxeo.template.rest`

{{#> callout type='info' }}
If you were extending the Nuxeo REST API, your package `org.nuxeo.ecm.restapi.server.jaxrs` must be updated to 
`org.nuxeo.ecm.restapi.server` to keep the Package Scan working.
{{/callout}}

### Nuxeo Component

> The following Nuxeo bundle names have changed in order to get rid of _JAX-RS_ occurrences:

Reviewer 4: ```suggestion
The following Nuxeo bundle names have changed in order to remove _JAX-RS_ occurrences:
```
